lib/smack/SmackWarnings.cpp

```diff
@@ -0,0 +1,95 @@
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/IR/DebugLoc.h"
+#include "llvm/IR/DebugInfoMetadata.h"
+
+#include "smack/SmackWarnings.h"
+#include "smack/SmackOptions.h"
+
+#include <algorithm>
+#include <set>
+
+namespace smack {
+using namespace llvm;
+
+std::string buildDebugInfo(const Instruction* i) {
+  std::stringstream ss;
+  if (i && i->getMetadata("dbg")) {
+    const DebugLoc DL = i->getDebugLoc();
+    auto* scope = cast<DIScope>(DL.getScope());
+
+    ss << scope->getFilename().str() << ":" << DL.getLine() << ":" << DL.getCol() << ": ";
+  }
+  return ss.str();
+}
+
+SmackWarnings::WarningLevel SmackWarnings::warningLevel = SmackWarnings::WarningLevel::Silent;
+
+
+void SmackWarnings::setWarningLevel(std::string warningTypeStr) {
+  if (warningTypeStr == "info")
+    warningLevel = WarningLevel::Info;
+  else if (warningTypeStr == "unsound")
+    warningLevel = WarningLevel::Unsound;
+  else if (warningTypeStr == "silent")
+    warningLevel = WarningLevel::Silent;
+  else
+    llvm_unreachable("invalid warning type.");
+}
+
+bool SmackWarnings::isSufficientWarningLevel(WarningLevel level) {
+  return warningLevel >= level;
+}
+
+SmackWarnings::UnsetFlagsT SmackWarnings::getUnsetFlags(RequiredFlagsT requiredFlags) {
```

Comment:
What does this do?

Reply:
It tests the list of required flags are set or not. For example, handling the half-precision intrinsics requires both `--float` and `--bit-precise`.  If one of the flags is set, we only need to remind the users to set the other.